zio-http/src/main/scala/zio/http/codec/TextChunkCodec.scala

```diff
@@ -0,0 +1,79 @@
+package zio.http.codec
+
+import scala.annotation.tailrec
+
+import zio.prelude._
+import zio.{Chunk, ChunkBuilder, NonEmptyChunk}
+
+sealed trait TextChunkCodec[A, I] {
```

Comment:
I like your idea, of having more detailed errors. But I still think the whole `TextChunkCodec` is not necessary.
I think you could just fit it in `TextCodec`. You could add two new methods to `TextCodec`, `def decodeAll(values: Chunk[String]): Chunk[A]` and `def encodeAll(values: Chunk[A]): Chunk[String]`.
For `Query`, add a parameter `cardinality: Cardinality`
```scala
sealed trait Cardinality
object Cardinality {
  case object ExactyOne extends Cardinality
  case object AtLeastOne extends Cardinality
  case object Many extends Cardinality
}
```
And set this according to the constructors.
Then just use the two new methods in `EncoderDecoder` and throw an exception if cardinality and the size of the chunk do not match.

This has imho the following benefits:
- no extra class
- every query codec is just one object instead of 2
- still having better error messages, without wrapping each result of decoding a query in `DecodeResult`

This seems to me better in terms of maintainability and allocations during decoding.

Reply:
You ask me to drop one class and replace with another one. :)
It won't decrease the number of classes in code and definitely increase complexity because we put more functionality in TextCodec unrelated to its main purpose.

TextCodec is good for encoding/decoding scalar values, this is it's speciality. TextChunkCodec do not have an intersection with it, and have another specialization. I believe it is good to follow to the single responcibility principle and do not pull all functionality in one class.

Reply:
You are right, that my idea still have one new data structure. The difference is, there are only singletons and it is analysable. What you have right now, can't be translated into docs. I can't take a look at an instance of `TextChunkCodec` and know the cardinality.
Also I don't see any problem with extending `TextCodc` by these methods, since they can be added to the trait and would be the same for all implementations.
As well I'd argue that it is still single responsibility. The responsibility to transform text to values and values to text.
@vigoo @jdegoes do you have some opinion / other idea(s)?

Reply:
For documentation it is possible to add a method "description" like in TextCodec or add another way of tagging. This is not an issue.

Reply:
@987Nabil I added the method `cardinality` to `TextChunkCodec`.

I can't agree that `TextChunkCodec` and `TextCodec` have the same responsibility. `TextCodec` solves the scalar transformation. One string - one value. `TextChunkCodec` solves the issue of cardinality: check amount of values and coerce it to expected cardinality. It uses `TextCodec` to encode each string scalar to a typed value.

Reply:
I offered my opinion. And I don't feel like this is my field of expertise for zio-http (yet). I'll neither ask you to change nor approve this PR but rather summon my fellow zio-http team members ༼ つ ◕◕ ༽つ